docs/lesson-plans/week-5.md

```diff
@@ -0,0 +1,639 @@
+>See more information on [overriding class or interface 
+>methods](https://www.tutorialspoint.com/java/java_overriding.htm).
+
+>Dependencies are external classes (in the case of Java and other Object-Oriented languages) or functions/methods that 
+>are imported and which your code depends on. In compiled languages and sometimes 
+>[linting](https://stackoverflow.com/questions/8503559/what-is-linting), your code cannot be complied without access to 
+>the dependencies (since it needs to know what functionality is provided). In the case of our 
+>backend code, we have dependencies on Vert.x, Mockito, JUnit, and a few other libraries.
+
+This is what our `PostsRouter` looked like initially:
+```java
+public class PostsRouter implements IRouter {
+    IPostsProcessor processor;
+
+    // The basic constructor.
+    public PostsRouter(IPostsProcessor processor) {
+        this.processor = processor;
+    }
+
+    @Override
+    public Router initializeRouter(Vertx vertx) {
+        Router router = Router.router(vertx);
+    ...
+```
+
+This is what our `PostsRouter` becomes with our externs:
+```java
+public class PostsRouter implements IRouter {
+    IPostsProcessor processor;
+    
+    // Our new externs.
+    Externals externs;
+
+    // The basic constructor.
+    public PostsRouter(IPostsProcessor processor) {
+        this.processor = processor;
+        this.externs = new Externals();
+    }
+
+    // The override constructor.
+    public PostsRouter(IPostsProcessor processor, Externals override) {
+        this.processor = processor;
+        this.externs = override;
+    }
+
+    public static class Externals {
+        public Router getRouter(Vertx vertx) {
+            return Router.router(vertx);
+        }
+    }
+
+    @Override
+    public Router initializeRouter(Vertx vertx) {
+        Router router = externs.getRouter(vertx);
+    ...
+```
+
+And this is how we override that behavior in the test class:
+```java
+public class PostsRouterTest {
+    ...
+
+    // Create the new externals, which overides the old behavior.
+    public static class TestExterns extends Externals {
+        public Router getRouter(Vertx vertx) {
+            return myTestRouter;
+        }
+    }
+
+    @BeforeEach
+    public void setup() {
+        ...
+        this.processor = new Processor(this.db, new TestExterns());
+        ...
+    }
+}
+```
+
+#### Exceptions
+
+When testing exceptions, JUnit provides methods and annotations to help, but we find it more helpful to use the 
+try/catch way of doing things. We think it allows you to test things about the exception better.
+
+Example:
+```java 
+try {
+    thingThatThrowsException();
+    fail(); // fail this test if we make it this far
+}
+catch (TheExpectedExceptionType e) {
+    assertEquals("My exception message", e.getMessage());
+    assertEquals(otherThingsAboutExceptionIfYouWant, e.doSomething());
+}
```

Comment:
Technically, this should also have a `catch (Exception e) {}` block in case an exception that we weren't expecting was thrown

Reply:
I think uncaught exceptions just fail the test automatically like how they're usually failed if something goes wrong